source/compiler/qsc_circuit/src/builder.rs

```diff
@@ -220,16 +237,55 @@ impl CircuitTracer {
         operations: &[OperationOrGroup],
         source_lookup: &impl SourceLookup,
     ) -> Circuit {
-        let operations = operations
+        let mut operations: Vec<Operation> = operations
             .iter()
             .map(|o| o.clone().into_operation(source_lookup))
             .collect();
+        let mut qubits = self.wire_map_builder.wire_map.to_qubits();
+        // We need to pass the original number of qubits, before any trimming, to finish the circuit below.
+        let num_qubits = qubits.len();
 
-        finish_circuit(
-            self.wire_map_builder.wire_map.to_qubits(),
-            operations,
-            source_lookup,
-        )
+        if self.config.prune_classical_qubits {
+            // Remove qubits that are always classical.
+            qubits.retain(|q| self.superposition_qubits.contains(&q.id));
+
+            // Remove operations that don't use any non-classical qubits.
+            operations.retain_mut(|op| self.should_keep_operation_mut(op));
+        }
+
+        finish_circuit(qubits, operations, num_qubits, source_lookup)
+    }
+
+    fn should_keep_operation_mut(&self, op: &mut Operation) -> bool {
```

Comment:
Can this operate on `OperationOrGroup` _before_ we transform it into `Operation`? `OperationOrGroup` is the "working" data structure which is more ergonomic to manipulate, and already has methods like `all_qubits()` . `Operation` is essentially the final "wire format" that we serialize to JSON, so we have to keep it stable but is kind of annoying to manipulate. You shouldn't have to change this function too much.